CONTRIBUTING.md

```diff
@@ -58,19 +58,130 @@ can be accepted.
+- To avoid unnecessary failures in GitHub Actions, make sure `make check` and `make test` work locally.
+See below for instructions to set up the development environment.
 
 ### Merging PRs
 
-- Before a PR is merged, all commits in the PR should be rebased into meaningful commits.
+- Before a PR is merged, all commits in the PR should meaningful commits.
   - Its commit message should be the same as the PR title if there only one commit.
-- PRs should be merged using the fast-forward only strategy.
-  - This ensures a linear commit history for the project.
-  - If the fast-forward merge does not work due to new commits on `main` (in this case the two branches diverge and the PR branch cannot be fast-forwarded), the PR branch should be rebased onto `main` first.
+- PRs should be merged using the `Squash and merge` strategy. In most cases a single commit with
+a detailed commit message body is preferred. Make sure to keep the `Signed-off-by` line in the body.
 
 ## Branching model
 
 * The `main` branch is only used for releases and the `staging` branch is used for development. We only merge to `main` when we want to create a new release for Macaron.
 
+## Setting up the development environment
+
+### Prerequisites
+
+- Python 3.11
+- Go 1.18
+- JDK 17
+
+### Prepare the environment
+
+To contribute to Macaron, clone the project and create a [virtual environment](https://docs.python.org/3/tutorial/venv.html) by using the [Makefile](https://www.gnu.org/software/make/manual/make.html#toc-An-Introduction-to-Makefiles):
+
+```bash
+make venv  # Create a new virtual environment in .venv folder using Python 3.11.
+```
+
+or for a specific version of Python:
+
+```bash
+PYTHON=python3.11 make venv  # Same virtual environment for a different Python version.
+```
+
+Activate the virtual environment:
+
+```bash
+. .venv/bin/activate
+```
+
+Finally, set up Macaron with all of its extras and initialize the local git hooks:
+
+```bash
+make setup
+```
+
+**Note**: Running the above command will prompt you for sudo access to install [Soufflé Datalog engine](https://github.com/souffle-lang/souffle). You can install Soufflé on your system before running `make setup` to avoid getting prompted.
+
+With that in place, you’re ready to build and contribute to Macaron!
+
+### Updating dependent packages
+
+It’s likely that during development you’ll add or update dependent packages in the `pyproject.toml` or `go.mod` files, which requires an update to the environment:
+
+```bash
+make upgrade
+```
+
+### Running Macaron as a Python package
+
+```bash
+usage: macaron [-h]
+```
+
+### Obtaining the GitHub personal access token
+Create your own Github access token (please refer to the instructions [here](https://docs.github.com/en/github/authenticating-to-github/keeping-your-account-and-data-secure/creating-a-personal-access-token)). When creating this token, make sure to assign **at least** `repo` permissions.
```

Comment:
I think we can update the instruction on permission of the Github token by using parts of the instructions we have for the Sphinx documentation

> Ideally, the GitHub token must have read permissions for the repositories that you want to analyze:
> 
> - Every [fine-grained personal-access token](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token#creating-a-fine-grained-personal-access-token) should have read permission to public GitHub repositories. However, if you are analyzing a private repository, please select it in the Repository Access section.
> - For [classic personal-access token](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token#creating-a-personal-access-token-classic), the repo.public_repo scope must be selected. Please select the whole repo scope if you are running the analysis against private repositories.
>

Reply:
See commit [6de60e8](https://github.com/oracle-samples/macaron/pull/279/commits/6de60e887c592f5ef3b4bb984985148507222b1f)